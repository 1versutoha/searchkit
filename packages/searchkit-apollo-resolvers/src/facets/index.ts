export { default as MultipleSelectFacet } from './MultipleSelectFacet'
export { default as RefinementSelectFacet } from './RefinementSelectFacet'
