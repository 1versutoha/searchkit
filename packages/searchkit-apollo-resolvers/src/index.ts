export * from './resolvers'
export { default as MultiMatchQuery } from './query/MultiMatchQuery'
