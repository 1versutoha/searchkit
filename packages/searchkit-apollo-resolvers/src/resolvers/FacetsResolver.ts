import SearchkitRequest from '../core/SearchkitRequest'
import FacetsManager from '../core/FacetsManager'

export default async (parent, {}, ctx) => {
  const { queryManager, config } = ctx
  const skRequest: SearchkitRequest = ctx.skRequest

  try {
    const fm = new FacetsManager(queryManager, config.facets)

    const aggs = fm.getAggregations()

    const results = await skRequest.search(aggs)

    const facets = fm.getFacetsFromResponse(results)

    return facets
  } catch (e) {
    console.log(e)
  }
}
