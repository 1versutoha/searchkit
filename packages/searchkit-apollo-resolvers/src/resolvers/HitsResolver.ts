import SearchkitRequest from '../core/SearchkitRequest'

interface HitsParameters {
  page: {
    start: number
    rows: number
  }
}

export default async (parent, parameters: HitsParameters, ctx: { skRequest: SearchkitRequest }) => {
  const { skRequest } = ctx

  try {
    const { hits } = await skRequest.search({
      from: parameters.page ? parameters.page.start : 0,
      size: parameters.page ? parameters.page.rows : 10
    })

    return hits.hits.map((hit) => ({
      id: hit._id,
      fields: hit._source
    }))
  } catch (e) {
    console.log(e)
  }
}
