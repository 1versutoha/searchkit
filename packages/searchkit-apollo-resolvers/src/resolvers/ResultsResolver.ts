import QueryManager, { FilterSet } from '../core/QueryManager'
import SearchkitRequest from '../core/SearchkitRequest'

export interface SearchkitConfig {
  host: string
  hits: {
    fields: string[]
  }
}

export interface ResultsResolverParameters {
  filters: Array<FilterSet>
  query: string
}

export default (config: SearchkitConfig) => async (parent, parameters, ctx) => {
  const queryManager = new QueryManager(parameters.filters, parameters.query)
  const skRequest = new SearchkitRequest(queryManager)

  ctx.skRequest = skRequest

  return {}
}
