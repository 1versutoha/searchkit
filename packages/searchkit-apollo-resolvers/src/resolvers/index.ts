export { default as HitsResolver } from './HitsResolver'
export { default as ResultsResolver, SearchkitConfig } from './ResultsResolver'
export { default as FacetsResolver } from './FacetsResolver'
