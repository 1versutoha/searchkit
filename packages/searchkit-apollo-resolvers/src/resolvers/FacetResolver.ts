import SearchkitRequest from '../core/SearchkitRequest'
import { getAggregationsFromFacets, getFacetsFromResponse } from '../core/FacetsFns'
import { SearchkitConfig } from './ResultsResolver'

interface FacetResolverParameters {
  id: string
  query: string
  page: {
    from: number
    size: number
  }
}

export default async (parent, parameters: FacetResolverParameters, ctx) => {
  const { queryManager } = ctx
  const config: SearchkitConfig = ctx.config
  const skRequest: SearchkitRequest = ctx.skRequest

  const facet = config.facets && config.facets.find((facet) => facet.getId())
  if (!facet) return null

  try {
    const aggs = getAggregationsFromFacets(queryManager, [facet])
    const results = await skRequest.search(aggs)
    const facets = getFacetsFromResponse([facet], results)

    return facets[0]
  } catch (e) {
    console.log(e)
  }
}
