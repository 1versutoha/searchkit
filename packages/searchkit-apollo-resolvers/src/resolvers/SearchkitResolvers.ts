import ResultsResolver, { SearchkitConfig } from './ResultsResolver'
import HitsResolver from './HitsResolver'
import FacetsResolver from './FacetsResolver'
import SummaryResolver from './SummaryResolver'

export default (config: SearchkitConfig) => ({
  ResultSet: {
    hits: HitsResolver,
    facets: FacetsResolver,
    // facet: FacetResolver,
    summary: SummaryResolver
  },
  Query: {
    results: ResultsResolver(config)
  },
  FacetSet: {
    __resolveType: (e) => e.type
  }
})
