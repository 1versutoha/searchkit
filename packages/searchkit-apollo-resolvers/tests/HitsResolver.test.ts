import { SearchkitConfig } from '../src/resolvers/ResultsResolver'
import { setupTestServer, callQuery } from './support/helper'
import { MultiMatchQuery } from '../src'

describe('Hits Resolver', () => {
  describe('should return as expected', () => {
    const runQuery = async (query = '') => {
      const gql = `
        {
          results(query: "${query}") {
            hits(page: {rows: 10, start: 0 }) {
              id
              fields {
                writers
                actors
              }
            }
          }
        }
      `
      const response = await callQuery({ gql })
      return response
    }

    it('should return correct Results', async () => {
      const config: SearchkitConfig = {
        host: 'http://localhost:9200/movies/_search',
        hits: {
          fields: ['actors', 'writers']
        },
        query: new MultiMatchQuery({ fields: ['actors', 'writers', 'title^4', 'plot'] })
      }

      setupTestServer(config)

      const response = await runQuery()
      expect(response.body.data).toMatchSnapshot()
      expect(response.status).toEqual(200)
    })
  })
})
