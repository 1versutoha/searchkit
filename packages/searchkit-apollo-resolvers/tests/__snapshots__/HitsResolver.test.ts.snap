// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hits Resolver should return as expected should return correct Results 1`] = `
Object {
  "results": Object {
    "hits": Array [
      Object {
        "fields": Object {
          "actors": Array [
            "Tim Robbins",
            "Morgan Freeman",
            "Bob Gunton",
            "William Sadler",
          ],
          "writers": Array [
            "Stephen King",
            "Frank Darabont",
          ],
        },
        "id": "tt0111161",
      },
      Object {
        "fields": Object {
          "actors": Array [
            "Marlon Brando",
            "Al Pacino",
            "James Caan",
            "Richard S. Castellano",
          ],
          "writers": Array [
            "Mario Puzo",
            "Francis Ford Coppola",
          ],
        },
        "id": "tt0068646",
      },
      Object {
        "fields": Object {
          "actors": Array [
            "Christian Bale",
            "Heath Ledger",
            "Aaron Eckhart",
            "Michael Caine",
          ],
          "writers": Array [
            "Jonathan Nolan",
            "Christopher Nolan",
            "David S. Goyer",
            "Bob Kane",
          ],
        },
        "id": "tt0468569",
      },
      Object {
        "fields": Object {
          "actors": Array [
            "Peter Dinklage",
            "Lena Headey",
            "Maisie Williams",
            "Emilia Clarke",
          ],
          "writers": Array [
            "David Benioff",
            "D.B. Weiss",
          ],
        },
        "id": "tt0944947",
      },
      Object {
        "fields": Object {
          "actors": Array [
            "Tim Roth",
            "Amanda Plummer",
            "Laura Lovelace",
            "John Travolta",
          ],
          "writers": Array [
            "Quentin Tarantino",
            "Roger Avary",
          ],
        },
        "id": "tt0110912",
      },
      Object {
        "fields": Object {
          "actors": Array [
            "Edward Norton",
            "Brad Pitt",
            "Helena Bonham Carter",
            "Meat Loaf",
          ],
          "writers": Array [
            "Chuck Palahniuk",
            "Jim Uhls",
          ],
        },
        "id": "tt0137523",
      },
      Object {
        "fields": Object {
          "actors": Array [
            "Noel Appleby",
            "Alexandra Astin",
            "Sean Astin",
            "David Aston",
          ],
          "writers": Array [
            "J.R.R. Tolkien",
            "Fran Walsh",
            "Philippa Boyens",
            "Peter Jackson",
          ],
        },
        "id": "tt0167260",
      },
      Object {
        "fields": Object {
          "actors": Array [
            "Leonardo DiCaprio",
            "Joseph Gordon-Levitt",
            "Ellen Page",
            "Tom Hardy",
          ],
          "writers": Array [
            "Christopher Nolan",
          ],
        },
        "id": "tt1375666",
      },
      Object {
        "fields": Object {
          "actors": Array [
            "Alan Howard",
            "Elijah Wood",
            "Noel Appleby",
            "Sean Astin",
          ],
          "writers": Array [
            "J.R.R. Tolkien",
            "Fran Walsh",
            "Philippa Boyens",
            "Peter Jackson",
          ],
        },
        "id": "tt0120737",
      },
      Object {
        "fields": Object {
          "actors": Array [
            "Keanu Reeves",
            "Laurence Fishburne",
            "Carrie-Anne Moss",
            "Hugo Weaving",
          ],
          "writers": Array [
            "Andy Wachowski",
            "Lana Wachowski",
          ],
        },
        "id": "tt0133093",
      },
    ],
  },
}
`;
